app/services/ab_test_bucket.rb

```diff
@@ -0,0 +1,61 @@
+class AbTestBucket
+  include ActiveModel::Model
+
+  attr_reader :buckets
+
+  validate :within_100_percent
+
+  def initialize(buckets: { default: 100 })
+    @buckets = buckets
+    raise 'invalid bucket data structure' unless valid_bucket_data_structure?
+    ensure_numeric_percentages
+  end
+
+  def bucket(discriminator = nil)
+    return :misconfigured unless valid?
+    return :default if discriminator.blank?
+
+    max_sha = (16 ** 64) - 1
+    user_value = Digest::SHA256.hexdigest(discriminator).to_i(16).to_f / max_sha * 100
```

Comment:
potentially this might make for a good helper method, so in the spirit of "don't stub what you don't own" we could stub `percent(discriminator)` to be whatever we need?

Reply:
Great call - done in cc18432c8